recipes/quickfix/all/conanfile.py

```diff
@@ -25,24 +36,87 @@ def _source_subfolder(self):
+        version = Version(str(self.settings.compiler.version))
+        if self.settings.compiler == "gcc" and (
+                str(self.settings.compiler.cppstd).find("98") != -1 or
+                (version <= "5.5" and self.settings.compiler.cppstd == "None")):
+            return "auto"
+
+        if self.settings.compiler == "clang" and (
+                (version <= "5.0" and self.settings.compiler.libcxx == "libstdc++" and
+                 self.settings.compiler.cppstd == "None")):
+            return "auto"
+
+        return self.default_options["unique_ptr"]
+
+    @property
+    def _shared_ptr(self):
+        # libc++ doesn't know of tr1.
+        if str(self.settings.compiler).find("clang") != -1:
+            if self.settings.compiler.libcxx == "libc++":
+                return "std"
+
+        # gcc default standard for version 5.5 is gnu++98. In the 6 series, it
+        # changed to gnu++11.
+        version = Version(str(self.settings.compiler.version))
+        if self.settings.compiler == "gcc" and (
+                str(self.settings.compiler.cppstd).find("98") != -1 or
+                (version <= "5.5" and self.settings.compiler.cppstd == "None")):
+            return "tr1"
+
+        if self.settings.compiler == "clang" and (
+                (version <= "5.0" and self.settings.compiler.libcxx == "libstdc++" and
+                 self.settings.compiler.cppstd == "None")):
+            return "tr1"
+
+        return self.default_options["shared_ptr"]
+
+    def _configure_cmake(self):
+        if not self._cmake:
+            self._cmake = CMake(self)
+            self._cmake.definitions["BUILD_SHARED_LIBS"] = self.options.shared
+            self._cmake.definitions["HAVE_SSL"] = self.options.with_ssl
+            self._cmake.definitions["HAVE_POSTGRESQL"] = self.options.with_postgres
+            self._cmake.definitions["SHARED_PTR"] = str(self.options.shared_ptr).upper()
+            self._cmake.definitions["UNIQUE_PTR"] = str(self.options.unique_ptr).upper()
+            self._cmake.configure(source_folder=self._source_subfolder, build_folder=self._build_subfolder)
+        return self._cmake
+
     def source(self):
         tools.get(**self.conan_data["sources"][self.version])
         os.rename(self.name + "-" + self.version, self._source_subfolder)
 
     def requirements(self):
-        if self.options.ssl:
+        if self.options.with_ssl:
             self.requires("openssl/1.1.1g")
 
+        if self.options.with_postgres:
+            self.requires("libpq/11.5")
+
     def config_options(self):
         if self.settings.os == "Windows":
             del self.options.fPIC
 
-    def _configure_cmake(self):
-        if not self._cmake:
-            self._cmake = CMake(self)
-            self._cmake.definitions["HAVE_SSL"] = self.options.ssl
-            self._cmake.configure(source_folder=self._source_subfolder)
-        return self._cmake
+        self.options.unique_ptr = self._unique_ptr
```

Comment:
```suggestion
```

You should not force an option, also unique_ptr it's available since C++11, which means it's largely available. You could force to unique_ptr only.

Reply:
Sorry, I've accepted this last suggestion of yours but I'm now reconsidering. Those `_unique_ptr` and `_shared_ptr` properties are intended to rule out invalid choice for those options. For instance, if the standard is set to C++17 or C++20, `auto` is not a valid option for `unique_ptr`.

As for your previous suggestion making these properties return from `self.options` and not from `self.default_options`, I think options are not forced unless they are invalid. Don't you agree?

Reply:
Also, for a number of settings, the compiler complains about the absence of `unique_ptr` in the `std` namespace. In such cases I've forced the option `unique_ptr` to `auto` which makes QuickFIX define `SmartPtr` as `std::auto_ptr`, which is a thing the library does and considers a valid option (as anachronistic as it may be).

Reply:
Maybe `shared_ptr` and `unique_ptr` shouldn't be options at all? This is an important issue since those options will alter the library's interface, as, for instance `SmartPtr` is a pre-compiler definition and appears, for instance here:

https://github.com/tarc/conan-center-index/blob/f780f5bc42717cda9b2a28bf82e11196f3084e4d/recipes/quickfix/all/test_package/executor.cpp#L13-L28

Reply:
> As for your previous suggestion making these properties return from self.options and not from self.default_options, I think options are not forced unless they are invalid. Don't you agree?

Default options is used for set the initial value for that specific option. You should not read from default_options, but from self.options only. Consider default_options as write-only.

Reply:
I see more problem related to the project, than to recipe. Did you communicate the upstream, or submitted a PR to them? Your patches changes a lot of stuff related to smart pointers and this recipe turned more complex.

Reply:
As for the patches concerning smart pointers, the only real change that was made to the code was to fix the condition upon which it chooses `std::tr1::shared_ptr` over `std::shared_ptr`:

https://github.com/conan-io/conan-center-index/blob/0309f2af11e49c54ebccfd026b0b425267f031a2/recipes/quickfix/all/patches/0002-fix-condition-for-using-tr1-namespace.patch#L16-L22

The other changes related to smart pointers are done to avoid the detection of support to those types during CMake generation. As the code itself is designed to be configurable in this respect, it's ok to just set the way `AutoPtr` and `ptr::shared_ptr` are to be implemented, without relying on the detection.

I thought that it would be better to avoid such a detection during package creation - I fear that could be some mismatch between them as detected by the package creator and the client itself due to something related to the way the toolchains are installed.

For instance, imagine that for some installation `std::unique_ptr` is ok in C++98 mode (maybe only issuing some warning), but in some other machine, even sharing the same options and settings, the compiler would report error when exposed to `std::unique_ptr` on C++98 mode (maybe some local change turn warnings into errors). Therefore a package created in the first machine would break compilation for clients using it in the second one.

So I thought it would be better to set the way `AutoPtr` and `ptr::shared` are defined in the recipe, based upon a case analysis on the compiler settings and some options.

Reply:
If you think it's better I can change back the way smart pointers are being dealt with.

Reply:
If it's not a project bug, so you have to change back. There other gaps using conan which can result in incompatibility, like glibc, but we don't deal with that, we ask to the users building from sources when necessary. It's the same case here, don't try to solve all problems, it will increase complexity to this recipe, will be hard to maintain, and will fix only a specific problem for an unique recipe.

Reply:
Hi, @uilianries . I changed back the way smart pointer are being dealt with.